File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;

class Solution {
    public int[] solution(int[] numbers) {
        int[] answer = new int[numbers.length];
        Deque<Integer> stack = new ArrayDeque<>();
        
        for (int i = 0; i < numbers.length; i++) {
            while (!stack.isEmpty()) {
                int idx = stack.peek();
                
                if (numbers[idx] < numbers[i]) {
                    answer[idx] = numbers[i];
                    stack.pop();
                } else {
                    break;
                }
            }
            
            stack.push(i);
        }
        
        while(!stack.isEmpty()) {
            answer[stack.pop()] = -1;
        }
        return answer;
    }
}

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

vector<int> solution(vector<int> numbers) {
   
    vector<int> answer (numbers.size(), -1);
    stack<pair<int, int>> s;
    
    for(int i = 0; i < numbers.size(); ++i)
    {
        while(!s.empty())
        {
            if (s.top().first < numbers[i])
            {
                auto& [num, idx] = s.top();
                s.pop();
                
                answer[idx] = numbers[i];
            }
            else
            {
                break;
            }
        }
        s.push({numbers[i], i});
    }
    
    return answer;
}